### PR TITLE
drivers/uart: stm32: Fix TC flag handling in IRQ driven mode

### LINKED ISSUE #38519: stm32: g0b1re: Log/Shell subsys with serial uart buggy after #38432
**Describe the bug**
The log and shell subsys is acting strangely (sluggish and buggy) after the merge of #38432. Most notably, the shell responds only after second keystrokes. For example if `a` is pressed, nothing appears on the screen. Pressing `a` again will have two `a`s on the shell.

I'm seeing this on my G0B1RE custom board, reverting #38432 solves the issue.

**To Reproduce**
I've tried the `hello_world` with `CONFIG_SHELL` enabled on the `nucleo_g0b1re`, same issue as my custom board

**Environment (please complete the following information):**
 - OS: Win10
 - Toolchain: GNUARMEMB
 - f83dc928f6

**Logs**
```
I: Starting bootloader
I: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x1
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Swap type: none
I: Bootloader chainload address offset: 0x10000
I: Jumping to the first image slot
***** delaying boot 2000ms (per build configuration) *****





uild time: Sep 15 2021 16:52:30 Malaysia!





1: Turn on modem

32m[00:00:04.011,000] <inf> hw_manager: Starting modem
gmoc:~ $
```

### LINKED ISSUE #38620: STM32 uart driver prevent system to go to deep sleep
**Describe the bug**
When using uart driver on STM32 (tested on `nucleo_wb55rg`) system can't go to sleep.

When printing logs (or printk) in the console using `uart_poll`, transfer complete isr is only called for the last byte of the string and not the previous ones.
In consequence `pm_constraint_release(PM_STATE_SUSPEND_TO_IDLE);` is not called and prevent system going to sleep.

**To Reproduce**
1. west build -b nucleo_wb55rg -s zephyr/samples/boards/stm32/power_mgmt/blinky/ -d build-pm
2. west flash
3. west debugserver

**Expected behavior**
Transfert complete isr is called for each bytes and `pm_constraint_release` is called.

**Impact**
System can't go to deep sleep state

**Logs and console output**
From gdb after printing some data (file `pm/pm_ctrl.c`):
```
p power_state_disable_count
$1 = {0, 0, 76, 0, 0, 0, 0}
```

**Environment (please complete the following information):**
 - OS: Ubuntu 20.04
 - Toolchain : zephyr-sdk.0.13.0
 - Commit SHA : 81595ed4db2081e227d6425acb25b6eec8191787

**Additional context**

### PR DESCRIPTION
**Pull request by erwango**
When modifying driver to use PM constraints to handle PM requests,
some change of behavior was introduced in IRQ driven mode,
to accommodate with the new usage of _TC flag.

This had impact in some scenarios, like shell for instance.
Likely, TC flag was cleared too soon, having consequences on
behavior on functions were it is checked (irq_is_pending,
irq_tx_complete).

Fix this by clearing the flag on fifo_fill.

EDIT:
~~Additionally, remove code enabling IT on driver init,
as this doesn't work well with some platforms (F4).~~ > Fixed by the following:

EDIT: Additionally, in polling mode take into account byte streams (Fixes #38620)

Tested on:
- nucleo_wb55rg (shell, uart_basic_api, pm, shell + pm)
- nucleo_f429zi (shell, uart_basic_api, uart_async_api)

Fixes #38519

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

EDIT: Also fixes:
- tests/drivers/uart/uart_async_api/drivers.uart.uart_async_api disco_l475_iot1

### CODE REVIEW

**Comment by erwango** [comment]
^^@ycsin

**Comment by gmarull** [comment]
@erwango Fixes should point to #38519

**Comment by ycsin** [comment]
I tried it briefly in my firmware, seems to solve the issue

File: drivers/serial/uart_stm32.c [hunk 1]
```diff
@@ -803,8 +805,6 @@ static void uart_stm32_isr(const struct device *dev)
 
 #ifdef CONFIG_UART_INTERRUPT_DRIVEN
 	if (LL_USART_IsActiveFlag_TC(UartInstance)) {
-		LL_USART_ClearFlag_TC(UartInstance);
```
Comment (gmarull) [inline drivers/serial/uart_stm32.c:806]: I've done some quick tests, seems that by just removing this line the shell works as before. It probably requires some more investigation, the IRQ API is really low-level some we've probably missed something.

**Comment by ycsin** [comment]
I test this more today and found that this is causing my g0b1 custom board to hard fault.

## More context
My custom board has g0b1 and EC21 gsm modem, they work over UART. Because the current stm32 uart doesnt has the implementation to wake the MCU from LPM, I have to keep the MCU awake while while the EC21 is operating.

I have a custom hardware manager implementation to handle this hardware dependencies and that is also using the `pm_constraint_*` functions.

**Comment by erwango** [comment]
> I revert to my own implementation in #38437 and the problem goes away.
@ycsin Thanks for reporting this. Would you mind re-opening #38437?

**Comment by erwango** [comment]
> It made the shell not sluggish, but there is still some garbage when the system starts:

Would you mind trying latest version of this PR ?

**Comment by ycsin** [comment]
> Would you mind trying latest version of this PR ?

Yeah, already tried, same issue

**Comment by erwango** [comment]
Also fixes:
- tests/drivers/uart/uart_async_api/drivers.uart.uart_async_api disco_l475_iot1

**Comment by ycsin** [comment]
Just tried the latest version, but nothing happens after bootloader jump, haven't use a hw debugger to trace though.

```
I: Starting bootloader
I: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x1
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Swap type: none
I: Bootloader chainload address offset: 0x10000
I: Jumping to the first image slot
```

**Comment by erwango** [comment]
> test version, but nothing happens after bootloader jump, haven't use a hw debug

@ycsin Thanks for the heads up but without more indications, it won't he able to fix the issue you're obesrving. Can you provide more context ? What is the target used ? What is the binary configuration (PM, IRQ , ....) ?

**Comment by ycsin** [comment]
@erwango 
Im using a g0b1re custom board

```conf
###################################################################
#################               Misc              #################
CONFIG_NEWLIB_LIBC=y
# Prevents overflow
CONFIG_IDLE_STACK_SIZE=1024
# Minimal Heap mem pool size for the Hawkbit working
CONFIG_HEAP_MEM_POOL_SIZE=4096
CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=4096
# Modem
CONFIG_TEST_RANDOM_GENERATOR=y
CONFIG_GSM_PPP_AUTOSTART=n

# CONFIG_SIZE_OPTIMIZATIONS=y

###################################################################
#################         Configures Boot         #################
CONFIG_BOOT_BANNER=n
CONFIG_BOOT_DELAY=2000
###################################################################
#################       Configures MCUBOOT        #################
CONFIG_BOOTLOADER_MCUBOOT=y
CONFIG_MCUBOOT_GENERATE_CONFIRMED_IMAGE=y
CONFIG_MCUBOOT_SIGNATURE_KEY_FILE=""
CONFIG_MCUBOOT_EXTRA_IMGTOOL_ARGS="--version 1.0.1"

###################################################################
#################         Configures Net          #################
# PPP networking support
CONFIG_NET_PPP=y
CONFIG_NET_L2_PPP=y
CONFIG_NET_NATIVE=y
CONFIG_NETWORKING=y

CONFIG_NET_L2_PPP_TIMEOUT=10000
# CONFIG_UART_MUX_RINGBUF_SIZE=1024
# CONFIG_NET_PPP_RINGBUF_SIZE=1024
CONFIG_UART_MUX_RINGBUF_SIZE=512
CONFIG_NET_PPP_RINGBUF_SIZE=512

# IPv4 enables PPP IPCP support
CONFIG_NET_IPV4=y
CONFIG_NET_IPV6=n

# Required for multiple MQTT clients and HTTP client to work
CONFIG_POSIX_MAX_FDS=8
CONFIG_NET_BUF_RX_COUNT=16
CONFIG_NET_BUF_TX_COUNT=32
CONFIG_NET_PKT_RX_COUNT=8
CONFIG_NET_PKT_TX_COUNT=8

CONFIG_NET_BUF_DATA_SIZE=256

###################################################################
#################          Configures PM          #################
CONFIG_PM=y
CONFIG_PM_DEVICE=y

###################################################################
#################      Configures Log subsys      #################

###################################################################
#################    Configures Setting subsys    #################
CONFIG_SETTINGS=y
CONFIG_SETTINGS_RUNTIME=y

###################################################################
#################     Configures Shell subsys     #################
CONFIG_SHELL=y
CONFIG_SHELL_MINIMAL=n
CONFIG_SHELL_VT100_COLORS=y
CONFIG_SHELL_PROMPT_UART=""
CONFIG_DATE_SHELL=n
CONFIG_KERNEL_SHELL=y
CONFIG_NET_SHELL=n
CONFIG_MODEM_SHELL=n
CONFIG_ADC_SHELL=n
CONFIG_I2C_SHELL=n
CONFIG_SENSOR_SHELL=n
CONFIG_FLASH_SHELL=n
```

**Comment by ycsin** [comment]
@ I have rebased to the tip of main branch and tested your latest pr 
9 min ago, as of: 
image

My app is now able to boot, with some garbage uart data at the beginning:

```
I: Starting bootloader
I: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x1
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Swap type: none
I: Bootloader chainload address offset: 0x10000
I: Jumping to the first image slot
***** delaying boot 2000ms (per build configuration) *****





uild time: Sep 20 2021 17:34:30 Malaysia!


32mgmoc:~$[00:Odometer value: 1.7955527 km
00:04.014,000] <inf> telemetry_mqtt: Init OK
[00:00:04.015,000] <inf> telemetry: Init OK
[00:00:04.515,000] <inf> timesync: Initialized
```

As a reference, this is what it would be, before the recent pm changes:
```
I: Starting bootloader
I: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x1
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Swap type: none
I: Bootloader chainload address offset: 0x10000
I: Jumping to the first image slot


gmoc:~ $ ***** delaying boot 2000ms (per build configuration) *****


Build time: Sep 20 2021 17:34:30 Malaysia!


[00:00Odometer value: 1.7955527 km
:04.014,000] <inf> telemetry_mqtt: Init OK
[00:00:04.014,000] <inf> telemetry: Init OK
[00:00:04.515,000] <inf> timesync: Initialized
```

**Comment by erwango** [comment]
@ycsin Thanks for the heads up. 
I'm not able to reproduce the garbage output on my end. Would you have any hint on the conditions required to reproduce ?
Also, FYI, I'm still working on the patch as there's an unexpected IRQ when `uart_stm32_configure` is run, at least on some targets (wb)

**Comment by erwango** [comment]
@ycsin, @bbilas Would you mind checking https://github.com/zephyrproject-rtos/zephyr/pull/38686?
I've attempted a full rework of the solution.

**Comment by vgeneves** [comment]
@erwango I just tested [#38686](https://github.com/zephyrproject-rtos/zephyr/pull/38686), and it fixes the problem on my side.

Serial console is good, `power_state_disable_count` is decremented correctly and power saving mode is called (measured consumption is a few µAmps).

**Pull request closed without merging**